### PR TITLE
feat(interceptors): migrate decorator handler to new hooks

### PR DESCRIPTION
**Pull request by metcoder95**
- **feat(interceptors): migrate decorator handler to new hooks**

## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: lib/handler/decorator-handler.js [hunk 1]
```diff
@@ -9,63 +10,58 @@ module.exports = class DecoratorHandler {
...
-    return this.#handler.onComplete?.(...args)
+    return this.#handler.onResponseEnd?.(...args)
   }
 
-  onBodySent (...args) {
-    assert(!this.#onCompleteCalled)
-    assert(!this.#onErrorCalled)
-
-    return this.#handler.onBodySent?.(...args)
+  onResponseError (...args) {
```
Comment (metcoder95) [inline lib/handler/decorator-handler.js:58]: I believe we should introduce a new hook to handle pre-request or in-flight request errors (e.g. network timeouts after connect, or runtime errors before connecting with remote peer); it feels quite odd handling these kind of errors through `onResponseError` when the response has not even being received

**Comment by metcoder95** [comment]
Will adjust the tests for the decorator

**Review by ronag — COMMENTED** [review]
I'm thinking whether we should just deprecate/remove this?

**Comment by metcoder95** [comment]
What exactly? The methods? Definitely. But we can aim to land this one to at least unlock the interceptors, and create a further PR to drop them

**Review by ronag — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**